server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java

```diff
@@ -0,0 +1,265 @@
+            public void onTimeout(TimeValue timeout) {
+                logger.info("timed out while waiting for removal of decommissioned nodes");
+                nodesRemovedListener.onFailure(
+                    new OpenSearchTimeoutException(
+                        "timed out [{}] while waiting for removal of decommissioned nodes [{}] to take effect",
+                        timeout.toString(),
+                        nodesToBeDecommissioned.toString()
+                    )
+                );
+            }
+        }, allDecommissionedNodesRemovedPredicate);
+    }
+
+    /**
+     * This method updates the status in the currently registered metadata.
+     * This method also validates the status with its previous state before executing the request
+     *
+     * @param decommissionStatus status to update decommission metadata with
+     * @param listener listener for response and failure
+     */
+    public void updateMetadataWithDecommissionStatus(DecommissionStatus decommissionStatus, ActionListener<DecommissionStatus> listener) {
+        clusterService.submitStateUpdateTask(decommissionStatus.status(), new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                Metadata metadata = currentState.metadata();
+                DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.custom(DecommissionAttributeMetadata.TYPE);
+                assert decommissionAttributeMetadata != null && decommissionAttributeMetadata.decommissionAttribute() != null;
+                // we need to update the status only when the previous stage is just behind the expected stage
+                // if the previous stage is already ahead of expected stage, we don't need to update the stage
+                // For failures, we update it no matter what
+                int previousStage = decommissionAttributeMetadata.status().stage();
+                int expectedStage = decommissionStatus.stage();
+                logger.info(
+                    "attempting to update current decommission status [{}] with expected status [{}]",
+                    decommissionAttributeMetadata.status().stage(),
+                    decommissionStatus
+                );
+                if (previousStage >= expectedStage) return currentState;
+                if (expectedStage - previousStage != 1 && !decommissionStatus.equals(DecommissionStatus.FAILED)) {
+                    throw new DecommissioningFailedException(
+                        decommissionAttributeMetadata.decommissionAttribute(),
+                        "invalid previous decommission status found while updating status"
+                    );
+                }
+                Metadata.Builder mdBuilder = Metadata.builder(metadata);
+                DecommissionAttributeMetadata newMetadata = decommissionAttributeMetadata.withUpdatedStatus(decommissionStatus);
```

Comment:
Can we move state transitions to `DecommissionAttributeMetadata#withUpdatedStatus `

Reply:
done